sdks/agent-sdk/src/core/Agent.ts

```diff
@@ -0,0 +1,128 @@
+import EventEmitter from "node:events";
+import type { ContentCodec } from "@xmtp/content-type-primitives";
+import {
+  Client,
+  type ClientOptions,
+  type DecodedMessage,
+} from "@xmtp/node-sdk";
+import { filter } from "@/utils/filter";
+import { AgentContext } from "./AgentContext";
+
+interface EventHandlerMap<ContentTypes> {
+  error: [error: Error];
+  message: [ctx: AgentContext<ContentTypes>];
+  start: [];
+  stop: [];
+}
+
+export interface AgentOptions<ContentTypes> {
+  client: Client<ContentTypes>;
+}
+
+export type AgentMiddleware<ContentTypes> = (
+  ctx: AgentContext<ContentTypes>,
+  next: () => Promise<void>,
+) => Promise<void>;
+
+export class Agent<ContentTypes> extends EventEmitter<
+  EventHandlerMap<ContentTypes>
+> {
+  readonly client: Client<ContentTypes>;
+  #middleware: AgentMiddleware<ContentTypes>[] = [];
+  #isListening = false;
+
+  constructor({ client }: AgentOptions<ContentTypes>) {
+    super();
+    this.client = client;
+  }
+
+  static async create<ContentCodecs extends ContentCodec[] = []>(
+    signer: Parameters<typeof Client.create>[0],
+    // Note: we need to omit this so that "Client.create" can correctly infer the codecs.
+    options?: Omit<ClientOptions, "codecs"> & { codecs?: ContentCodecs },
+  ) {
+    const client = await Client.create(signer, options);
+    return new Agent({ client });
+  }
+
+  static async build<ContentCodecs extends ContentCodec[] = []>(
+    identifier: Parameters<typeof Client.build>[0],
+    // Note: we need to omit this so that "Client.build" can correctly infer the codecs.
+    options?: Omit<ClientOptions, "codecs"> & { codecs?: ContentCodecs },
+  ) {
+    const client = await Client.build(identifier, options);
+    return new Agent({ client });
+  }
+
+  use(middleware: AgentMiddleware<ContentTypes>) {
+    this.#middleware.push(middleware);
+    return this;
+  }
+
+  async start() {
+    if (this.#isListening) {
+      return;
+    }
+
+    try {
+      this.#isListening = true;
+      void this.emit("start");
+
+      const stream = await this.client.conversations.streamAllMessages();
+      for await (const message of stream) {
+        // The "stop()" method sets "isListening"
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        if (!this.#isListening) break;
+        try {
+          await this.processMessage(message);
+        } catch (error) {
+          this.throwError(error);
+        }
+      }
+    } catch (error) {
+      this.#isListening = false;
+      this.throwError(error);
+    }
+  }
+
+  private async processMessage(message: DecodedMessage<ContentTypes>) {
```

Comment:
can we use `#` here instead? `async #processMessage(message: DecodedMessage<ContentTypes>) {`